src/device/dnomem.c

```diff
@@ -0,0 +1,257 @@
+	NOMEM_ACCESS_ENTER_DEBUGGER,
+	NOMEM_ACCESS_HALT_MACHINE,
+};
+
+/** Device no-mem instance data structure */
+typedef struct {
+	/** No-memory location. */
+	ptr36_t addr;
+	/** No-memory size. */
+	uint64_t size;
+	/** No-memory behavior on access. */
+	enum nomem_access_handling_e mode;
+} dnomem_data_s;
+
+
+static const char *nomem_access_handling_as_str(enum nomem_access_handling_e mode)
+{
+	switch (mode) {
+	case NOMEM_ACCESS_WARN:
+		return "warn";
+	case NOMEM_ACCESS_ENTER_DEBUGGER:
+		return "debug";
+	case NOMEM_ACCESS_HALT_MACHINE:
+		return "halt";
+	}
+	return "unknown";
+}
+
+/** Init command implementation
+ *
+ * @param parm Command-line parameters
+ * @param dev  Device instance structure
+ *
+ * @return True if successful
+ *
+ */
+static bool dnomem_init(token_t *parm, device_t *dev)
+{
+	parm_next(&parm);
+	uint64_t start_addr = parm_uint_next(&parm);
+	uint64_t size = parm_uint_next(&parm);
+
+	if (!phys_range(start_addr)) {
+		error("Physical memory address out of range");
+		return false;
+	}
+
+	if (!phys_range(start_addr + size)) {
+		error("Invalid address, size would exceed the physical "
+		    "memory range");
+		return false;
+	}
+
+	if (!ptr36_dword_aligned(start_addr)) {
+		error("Physical memory address must be 8-byte aligned");
+		return false;
+	}
+
+	/* Allocate structure */
+	dnomem_data_s *data = safe_malloc_t(dnomem_data_s);
+	dev->data = data;
+
+	/* Basic structure inicialization */
+	data->addr = start_addr;
+	data->size = size;
+	data->mode = NOMEM_ACCESS_WARN;
+
+	return true;
+}
+
+/** mode command implementation.
+ *
+ * @param parm Command-line parameters
+ * @param dev  Device instance structure
+ *
+ * @return True if successful
+ */
+static bool dnomem_setmode(token_t *parm, device_t *dev)
+{
+	dnomem_data_s *data = (dnomem_data_s *) dev->data;
+	const char *const mode = parm_str(parm);
+
+	if (strcmp(mode, "warn") == 0) {
+		data->mode = NOMEM_ACCESS_WARN;
+	} else if (strcmp(mode, "debug") == 0) {
+		data->mode = NOMEM_ACCESS_ENTER_DEBUGGER;
+	} else if (strcmp(mode, "halt") == 0) {
+		data->mode = NOMEM_ACCESS_HALT_MACHINE;
+	} else {
+		error("Unknown mode %s, expecting warn, debug or halt.", mode);
+		return false;
+	}
+
+	return true;
+}
+
+/** Info command implementation
+ *
+ * @param parm Command-line parameters
+ * @param dev  Device instance structure
+ *
+ * @return True (always successful)
+ *
+ */
+static bool dnomem_info(token_t *parm, device_t *dev)
+{
+	dnomem_data_s *data = (dnomem_data_s *) dev->data;
+
+	printf("[address  ] [size     ] [mode]\n"
+	    "%#011" PRIx64 " %#011" PRIx64 " %s\n",
+	    data->addr, data->size, nomem_access_handling_as_str(data->mode));
+
+	return true;
+}
+
+
+/** Dispose disk
+ *
+ * @param dev Device pointer
+ *
+ */
+static void dnomem_done(device_t *dev) {
+	safe_free(dev->data);
+}
+
+static void dnomem_access32(const char *operation_name, device_t *dev, ptr36_t addr)
+{
+	dnomem_data_s *data = (dnomem_data_s *) dev->data;
```

Comment:
I would probably unswitch this and just store a pointer to a handler depending on the configuration (strategy pattern :-), so that there is no need to switch on every memory access.

Reply:
Done.